ciao-cli/main.go

```diff
@@ -855,127 +295,96 @@ func checkCompulsoryOptions() {
 		fatal += "Missing required Ciao controller URL\n"
 	}
 
-	if *allInstances == true && *instance != "" {
-		fatal += "All instances or one single instance ?\n"
-	}
-
 	if fatal != "" {
 		fatalf(fatal)
 	}
 }
 
-func cliList() {
-	if *listInstances == true {
-		listAllInstances(*tenantID, "", *instanceMarker, *instanceOffset, *listLength)
-	}
-
-	if *listWlInstances == true {
-		listAllInstances("", *workload, *instanceMarker, *instanceOffset, *listLength)
-	}
-
-	if *listCNInstances == true {
-		listNodeInstances(*computeNode)
-	}
-
-	if *listQuotas == true {
-		listTenantQuotas(*tenantID)
-	}
-
-	if *listResources == true {
-		listTenantResources(*tenantID)
-	}
-
-	if *listWorkloads == true {
-		listTenantWorkloads(*tenantID)
-	}
-
-	if *listComputeNodes == true {
-		listAllComputeNodes()
-	}
+func main() {
+	var err error
 
-	if *listCNCIs == true {
-		listAllCNCIs()
-	}
+	flag.Usage = usage
+	flag.Parse()
 
-	if *listLabels == true {
-		listAllLabels()
-	}
+	getCiaoEnvVariables()
+	checkCompulsoryOptions()
 
-	if *listEvents == true || *listAllEvents == true {
-		listClusterEvents(*tenantID, *listAllEvents)
+	/* If we're missing the tenant name let's try to fetch one */
+	if *tenantName == "" {
+		*tenantName, *tenantID, err = getTenant(*identityUser, *identityPassword, *tenantID)
+		if err != nil {
+			fatalf(err.Error())
+		}
+		warningf("Unspecified scope, using (%s, %s)", *tenantName, *tenantID)
 	}
-}
 
-func cliDump() {
-	if *clusterStatus == true {
-		dumpClusterStatus()
+	scopedToken, *tenantID, _, err = getScopedToken(*identityUser, *identityPassword, *tenantName)
+	if err != nil {
+		fatalf(err.Error())
 	}
 
-	if *dumpCNCI == true {
-		dumpCNCIDetails(*cnci)
+	// Print usage if no arguments are given
+	args := flag.Args()
+	if len(args) < 1 {
+		usage()
 	}
 
-	if *dumpLabel != "" {
-		dumpTraceData(*dumpLabel)
+	// Find item in cmdline args
```

Comment:
So our argument parsing needs to be a more complex than this.
We want to be able to support commands, sub-commands, and options (general and subcommand specific ones) in any particular order, except that a sub-command must come after a command. So those 2 are correct:
`ciao-cli -v=2 -identity=alberto node list -cnci`
`ciao-cli node list -cnci -v=2 -identity=alberto`

For us to handle those properly and not having to force users into following a fairly complex order, we need to do 3 things:

1) Implement a helper to separate options from commands:
```
func splitArgs(args []string) (options []string, commands []string) {
	for _, a := range args {
		if len(a) > 1 && a[0] == '-' {
			options = append(options, a)
		} else {
			commands = append(commands, a)
		}
	}

	return options, commands
}
```
2) Implement a helper to generate a new FlagSet, that will automatically include all the generic options:

```
func newFlagSet() (*flag.FlagSet) {
	f := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
	
	identityURL      = f.String("identity", "", "Keystone URL")
	identityUser     = f.String("username", "", "Openstack Service Username")
	identityPassword = f.String("password", "", "Openstack Service Password")
	controllerURL    = f.String("controller", "", "Controller URL")
	tenantID         = f.String("tenant-id", "", "Tenant UUID")
	tenantName       = f.String("tenant-name", "", "Tenant name")
        computePort      = f.Int("computeport", openstackComputePort, "Openstack Compute API port")

	return f
}
```

and main.go will contain:
```
var (
	identityURL      *string
	identityUser     *string
	identityPassword *string
	controllerURL    *string
	tenantID         *string
	tenantName       *string
        computePort       *int
)
```

3) from main.go we would
    a) call `options, cmds := splitArgs(os.Args)`
    b) `command := commands[cmds[0]]` -> that will give you the command implementation
    c) `command.ParseArgs()` where `ParseArgs()` for a command will:
         * Find the sub-command
         * create a new FlagSet: `newFlagSet()`
         * Add all sub-command specific options to it
         * Call Parse() on this new flagset.
     d) Call `getCiaoEnvVariables()` and `checkCompulsoryOptions`
     e) Call the 
```
	/* If we're missing the tenant name let's try to fetch one */
	if *tenantName == "" {
		*tenantName, *tenantID, err = getTenant(*identityUser, *identityPassword, *tenantID)
		if err != nil {
			fatalf(err.Error())
		}
		warningf("Unspecified scope, using (%s, %s)", *tenantName, *tenantID)
	}

	scopedToken, *tenantID, _, err = getScopedToken(*identityUser, *identityPassword, *tenantName)
	if err != nil {
		fatalf(err.Error())
	}
```
       block
     f) Call command.Run()